src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs

```diff
@@ -69,102 +67,79 @@ private unsafe static long GetAsciiStringAsLong(string str)
+        private unsafe static string GetAsciiString(MemoryPoolBlock2 start, MemoryPoolIterator2 end, int inputOffset, int length)
         {
-            // avoid declaring other local vars, or doing work with stackalloc
-            // to prevent the .locals init cil flag , see: https://github.com/dotnet/coreclr/issues/1279
-            char* output = stackalloc char[length];
-
-            return GetAsciiStringImplementation(output, input, inputOffset, length);
-        }
+            var asciiString = new string('\0', length);
 
-        private static unsafe string GetAsciiStringImplementation(char* output, byte[] input, int inputOffset, int length)
-        {
-            for (var i = 0; i < length; i++)
+            fixed (char* outputStart = asciiString)
             {
-                output[i] = (char)input[inputOffset + i];
-            }
-
-            return new string(output, 0, length);
-        }
-
-        private static unsafe string GetAsciiStringStack(MemoryPoolBlock2 start, MemoryPoolIterator2 end, int inputOffset, int length)
-        {
-            // avoid declaring other local vars, or doing work with stackalloc
-            // to prevent the .locals init cil flag , see: https://github.com/dotnet/coreclr/issues/1279
-            char* output = stackalloc char[length];
-
-            return GetAsciiStringImplementation(output, start, end, inputOffset, length);
-        }
-
-        private unsafe static string GetAsciiStringHeap(MemoryPoolBlock2 start, MemoryPoolIterator2 end, int inputOffset, int length)
-        {
-            var buffer = new char[length];
-
-            fixed (char* output = buffer)
-            {
-                return GetAsciiStringImplementation(output, start, end, inputOffset, length);
-            }
-        }
-
-        private static unsafe string GetAsciiStringImplementation(char* output, MemoryPoolBlock2 start, MemoryPoolIterator2 end, int inputOffset, int length)
-        {
-            var outputOffset = 0;
-            var block = start;
-            var remaining = length;
-
-            var endBlock = end.Block;
-            var endIndex = end.Index;
+                var output = outputStart;
+                var block = start;
+                var remaining = length;
 
-            while (true)
-            {
-                int following = (block != endBlock ? block.End : endIndex) - inputOffset;
+                var endBlock = end.Block;
+                var endIndex = end.Index;
 
-                if (following > 0)
+                while (true)
                 {
-                    var input = block.Array;
-                    for (var i = 0; i < following; i++)
+                    int following = (block != endBlock ? block.End : endIndex) - inputOffset;
+
+                    if (following > 0)
                     {
-                        output[i + outputOffset] = (char)input[i + inputOffset];
```

Comment:
Also If I didn't unroll I think this would be a perf regression as the string is initialized; rather than being operations on initialized stack memory; then a memcopy to an uninitialized string; however I think the code being less weird makes up for that...